File: Java/Sorting/Sorting.java
import java.io.*;
import java.util.*;

class Sorting {

    public static void merge(int arr[], int l, int m, int h) {
        int n1 = m - l + 1;
        int n2 = h - m;

        // create temp arrays
        int L[] = new int[n1];
        int R[] = new int[n2];

        // copy data to temp arrays
        for (int i = 0; i < n1; i++) {
            L[i] = arr[l + i];
        }

        for (int j = 0; j < n2; j++) {
            R[j] = arr[m + 1 + j];
        }

        // Merge temp arrays
        int i = 0, j = 0;

        // Initialize index of merged subrray array
        int k = l;

        // sort the arrays
        while (i < n1 && j < n2) {
            if (L[i] < R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            k++;
        }
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    public static void sort(int arr[], int l, int r) {
        if (l < r) {
            // Find the middle point
            int m = l + (r - l) / 2;

            // sort first and second element
            sort(arr, l, m);
            sort(arr, m + 1, r);
            // Merge the sorted halves
            merge(arr, l, m, r);
        }
    }

    public static int[] productExceptSelf(int[] arr) {
        int n = arr.length;
        int[] left = new int[n];
        int[] right = new int[n];
        int[] prod = new int[n];

        left[0] = 1;
        right[n - 1] = 1;

        for (int i = 1; i < n; i++) {
            left[i] = left[i - 1] * arr[i - 1];
        }

        // fill the right side of the array
        for (int j = n - 2; j > -1; j--) {
            right[j] = right[j + 1] * arr[j + 1];
        }

        // fill the prod array
        for (int i = 0; i < n; i++) {
            prod[i] = right[i] * left[i];
        }

        return prod;

    }

    public static void main(String args[]) {
        int arr[] = { 12, 11, 13, 5, 6, 7 };
        int[] nums = { 1, 2, 3, 4 };

        int[] ans = productExceptSelf(nums);

        System.out.println(Arrays.toString(ans));

    }
}

File: cpp/sorting/Sorting.cpp
#include <iostream> // preprocessor directive - before compilation all the files would be copeid
#include <vector>

using namespace std;

class Sort
{
public:
    void merge(int nums[], int left, int mid, int right)
    {
        int const n1 = mid - left + 1;
        int const n2 = right - mid;

        auto *leftArray = new int[n1], *rightArray = new int[n2];

        // Copy data to temp arrays leftArray[] and rightArray[]
        for (auto i = 0; i < n1; i++)
        {
            leftArray[i] = nums[left + i];
        }
        for (auto j = 0; j < n2; j++)
        {
            rightArray[j] = nums[mid + 1 + j];
        }

        auto leftArrIndex = 0,
             rightArrIndex = 0;
        int mergedArrIndex = left;
        while (leftArrIndex < n1 && rightArrIndex < n2)
        {
            if (leftArray[leftArrIndex] < rightArray[rightArrIndex])
            {
                nums[mergedArrIndex] = leftArray[leftArrIndex];
                leftArrIndex++;
            }
            else
            {
                nums[mergedArrIndex] = rightArray[rightArrIndex];
                rightArrIndex++;
            }
            mergedArrIndex++;
        }
    }

    void merge_sort(int nums[], int start, int end)
    {
        if (start < end)
        {
            int mid = start + (end - start) / 2;
            merge_sort(nums, start, mid);
            merge_sort(nums, mid + 1, end);

            merge(nums, start, mid, end);
        }
    }

    int partition(vector<int> &nums, int low, int high)
    {
        int pivot = nums[high];
        int i = low - 1;

        for (int j = low; j <= high - 1; j++)
        {
            if (pivot > nums[j])
            {
                i++;
                swap(nums, i, j);
            }
        }

        swap(nums, i + 1, high);
        return (i + 1);
    }

    void quickSort(vector<int> &nums, int low, int high)
    {
        if (low < high)
        {
            int pi = partition(nums, low, high);
            quickSort(nums, low, pi - 1);
            quickSort(nums, pi + 1, high);
        }
    }
    void swap(vector<int> &nums, int i, int j)
    {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
};

int main()
{
    vector<int> arr = {5, 4, 3, -10, 2, 1, 100};

    int nums[] = {5, 4, 3, -10, 2, 1, 100};
    Sort srt;
    srt.quickSort(arr, 0, arr.size() - 1);

    for (int i = 0; i < arr.size(); i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 1;
}
